### DISCUSSION TITLE
Belongs to Many: Prevent Attaching already Attached Model

### DISCUSSION THREAD

**Discussion by intrepidws** [filamentphp/filament#1510] [Help]
I have a Belongs to Many relationship between Registrants and Codes. I have a `CodesRelationManager` class (which extends `BelongsToManyRelationManager`) and am using that to allow codes to be attached to registrants. Currently, there is nothing to prevent someone from attaching a code to a registrant if that code is already attached to the registrant. Is there a way to prevent this?

I have the following code for my `attachForm` method:

```
public static function attachForm(Form $form): Form
{
    return $form
        ->schema([
            static::getAttachFormRecordSelect()
                ->searchable(['code', 'type', 'title']),
            Forms\Components\Select::make('quantity')
                ->required()
                ->options(array_combine(range(1, 100), range(1, 100)))
                ->default(1),
        ]);
}
```

**Comment by intrepidws** [comment]
I think that I found an actual bug surrounding this functionality. Currently Filament allows me to attach the same model multiple times. However, if I try to detach one of those multiple models - it detaches ALL of them. And the UI isn't aware that all were dropped, so it keeps showing all others unless I reload the page.

Back to the general issue of attaching duplicate models. I can think of some rare scenarios where you might want to attach the same model multiple times (and have different pivot table data for each) - in fact I have at least one project where I do this. Maybe the answer here is to prevent attaching duplicates by default, as I think that's the far more common scenario - and then have an `allowDuplicates()` method for those edge cases.

Anyway, this is what I was able to do to prevent attaching duplicate models. Perhaps this might be useful to others, or could be extrapolated into making this the default behavior.

```
public static function attachForm(Form $form): Form
{
    return $form
        ->schema([
            static::getAttachFormRecordSelect()
                ->searchable(['code', 'type', 'title'])
                ->rules([
                    function (Livewire $livewire) {
                        $registrant = $livewire->ownerRecord;

                        return function (string $attribute, $value, \Closure $fail) use ($registrant) {
                            if ($registrant->codes->contains($value)) {
                                $fail("This code already exists on this registrant record.");
                            }
                        };
                    },
                ]),
            Forms\Components\Select::make('quantity')
                ->required()
                ->options(array_combine(range(1, 100), range(1, 100)))
                ->default(1),
            Forms\Components\TextInput::make('other')
                ->helperText('This should only be used to provide a value for an "Other" selection for a demographic question.'),
        ]);
}

**Reply by danharrin** [reply]
Hm yeah that detach thing is a bug... not sure how to fix tbh.

**Reply by intrepidws** [reply]
I was thinking about this more and I'm not entirely sure it _is_ a bug. Unless the pivot table has an auto-incrementing ID, then the system doesn't have enough information to delete a specific item. That said, if the pivot table does not have an auto-incrementing ID - then the system shouldn't allow a duplicate attach in the first place.

**Reply by danharrin** [reply]
I think that, for simplicity, we should just enforce that people use an auto-incrementing ID for the pivot.

**Reply by Tklaversma** [reply]
@danharrin @intrepidws 

Old thread, but just for other people to know.

Lets say we have `facilities` that can be used by `accommodations` and `units`, in other words a `Many-to-Many polymorphic relationship`. In case of such a relationship, you always need 3 components:

1. ID of the table itself (primary auto-increment)
2. The foreign key to the `facilities` table
3. The morphs (facilityable_type` and `facilityable_id` columns with a compound index

So in Laravel you can do:
```php
Schema::create('facilityables', function (Blueprint $table) {
    $table->id();
    $table->foreignId('facility_id')->constrained();
    $table->morphs('facilityable');
});
```

Cheers 🍻
